include/zephyr/sys/util.h

```diff
@@ -26,6 +26,88 @@
 #include <zephyr/types.h>
 #include <stddef.h>
 
+/** @brief Number of bits that make up a type */
+#define NUM_BITS(t) (sizeof(t) * 8)
+
+#ifdef __cplusplus
+template <typename T> static inline constexpr int __z_log2_impl(T x)
+{
+	if (sizeof(x) <= sizeof(unsigned int)) {
+		return (x == 0) * (-1) + (x != 0) * (NUM_BITS(unsigned int) - __builtin_clz(x) - 1);
+	} else if (sizeof(x) <= sizeof(unsigned long)) {
+		return (x == 0) * (-1) +
+		       (x != 0) * (NUM_BITS(unsigned long) - __builtin_clzl(x) - 1);
+	} else if (sizeof(x) <= sizeof(unsigned long)) {
+		return (x == 0) * (-1) +
+		       (x != 0) * (NUM_BITS(unsigned long long) - __builtin_clzll(x) - 1);
+	}
+
+	__ASSERT_NO_MSG(sizeof(x) == sizeof(unsigned long long));
+}
+
+template <typename T> static inline constexpr int __z_log2ceil_impl(T x)
+{
+	if (sizeof(x) <= sizeof(unsigned int)) {
+		return (x > 1) * NUM_BITS(unsigned int) - __builtin_clz(x - 1);
+	} else if (sizeof(x) <= sizeof(unsigned long)) {
+		return (x > 1) * NUM_BITS(unsigned long) - __builtin_clzl(x - 1);
+	} else {
+		return (x > 1) * NUM_BITS(unsigned long long) - __builtin_clzll(x - 1);
+	}
+
+	__ASSERT_NO_MSG(sizeof(x) == sizeof(unsigned long long));
```

Comment:
What does this assert do? It looks to be unreachable, maybe add a comment or message to explain.

Reply:
Yea, it was to assert against an unreachable piece of code. I ended up using C++ `static_assert()` because Zephyr's `__ASSERT_NO_MSG()` is not declared and I wasn't sure if it made sense to pull in `<zephyr/sys/__assert.h>` here.